### PR TITLE
Azur API Client Rework

### PR DESCRIPTION
**Pull request by Tsukugi**
Here is a summary of all of this PR changes

## TL:DR

Everything has changed, there is two clients now (Hiei and AzurAPI), both work nicely with type inference, and there is support for both Node and Browser.

### Why?

While trying locally, I noted that types were incorrectly exported from the client, for example, `instance.apis.ships.get()` would show a TS error as internally the client assign `ships: Ships | Hiei.Ships` which they have totally diferent sets of methods.
So my first approach was to use type guards to workaround this issue, which didnt work. 

The second issue that I found was that this only works on node! (>14 to be exact). This was a problem for me as I wanted to try this API for a browser environment, so I started to think for a way to achieve this. I take blame on this a bit, as I didn't notice it beforehand.

This started as a simple check on why this was not working, and ended up since last week tinkering around how to improve this package, of course I made this because it's fun and I really like Azur Lane. 

### What has changed?

As the title says, I reworked the client, api, updater and the cache almost from scratch, while still being faithful as much as possible on how the current Client was being used externally.

The project is separated in the following modules:

1. Client: The client has the functionality of setting everything before being used. But with this issue on the API's exported types, I decided to separate HieiAPI and Local/Azur API into separate clients. Each has a class in which the user can safely use and have the correct API typing. 

    Now the clients have two ways to be used. We have the Classes `AzurAPI` and `HieiAzurAPI` (Legacy) which are still as simple, easy to use, instantiate and done, kind of Classes. And we have now the `createClient` and `createHieiClient` which are what lies behind them, which acomplish the same thing, just with a diferent syntax. Also there is now a `createClientFactory` which also is the client maker for both, so it's possible to create another client from this.

    Regarding Hiei, Hiei is a very simple API compared to the AzurAPI which has Updater, Cache and Events, so Hiei has now a really simple client exported, and only AzurAPI client uses all of these modules.

2. API: These are the most similar as before, I mainly replaced it with function sets that do actions depending on API provided `state` argument. Currently the API works both as a API and cacheHandler which may be fine, but it also depends on the Client method `set()` which would be a bit weird, considering that logically speaking, the Client depends on the API, but the API also depends on the Client, causing them to be deeply coupled.  

    Regarding the AzurAPI, some structures were changed, API Class is removed and instead all the shared logic is on SharedAPI, which is used by the AzurAPI to do the complex logic, the Azur APIs define a custom `fuse` function if needed (Mainly because of the differences of the API and their own keys to search with).

    The biggest change here is that the API now only runs the different methods on the provided data and returns an answer, so no caching nor data modification.

2. State: Another quite big change is that instead of caching we are going to use a really simple state machine, that works a bit similar like Saratoga (I was really further on this until I found out about Sara). 

    This mainly get the data as `Dataset` which is an object that offers an array (mostly used) and a dict (for direct lookup by id without needing a search function, this does not replace SearchAPI as fuse is generally more powerful with the fuzzy search).

3.  ClientTools (Events, FileManager, Updater, Fetch and LocalFiles): Here lied most of the Node dependant codebase, there are currently 2 main changes:

    First, to address the main issue (Node and Browser compatibility), I define and use templates to avoid having to check every call if it Node or Browser. The templates (e.g `UpdaterTemplate`, `EventsTemplate`) are objects that offer generic functions to be called on the inner code. So the Updater would call, for example, `events.emit(Events.debug, "Hi from a debug!")`. But there is no real implementation of `events.emit` on updater. But instead, the client (on `createClient`, using `getClientTools`) will define everything on the initialization of the client.

    Second, the lifecycle on the client works like this: Initialization of structures -> `createClient` starts EventManager and then Updater  -> Return the instance to the user. While asynchronously Updater will do: (using `versionHandler`) ask for latest version of everything, if not, then request -> Will register the fetched data into the `State` and `LocalFiles`. (On browser, file managing is awful so it's not supported at the moment, but a client will do everything else).

#### General Changes

Here I wanted to achieve 2 things: Standardization and use Functional programming.

For Standardization, I use `prettier` to have a even codebase in the project. This is available with `npm run prettier` commands. Prettier will force all the codebase to be to one predefined code style.
 
As for Functional Programming, this mainly applies to the API, but also is applied in the rest of the project. Having pure functions that dont modify anything external, just do something simple and return an answer is may more maintainable than having huge and complex module instances, this also brings another benefit, that everything is quite decoupled so it can be reused for anything new. 

#### Webpack? 

Webpack is used to build a umd esm bundle that generally is used on browser (I define another tsconfig.json to achieve this), and `tsc` still builds node modules as `commonjs` to suport node. So at the end we have the same build files with one extra `azurapi.umd.bundle.js` if you use ESM import, otherwise (on require) will use `index.js`

Thanks for reading if you really read until here!
Feel free to ask any questions on discord if needed

Tsukugi

### CODE REVIEW

**Comment by Deivu** [comment]
Hiei is depracated in favor of Formidable; (Docs: https://github.com/Deivu/Formidable#api-ratelimits), if you haven't changed Hiei to Formi, please do change it.

We now also have a public hosted api for that which has a base url of https://formidable.kashima.moe/
The return data should follow the per-spec of the raw data, and will always be on per-spec of the raw data.

**Comment by jasperchua99** [comment]
due to inactivity over time... might wanna close this, can be reopened in the future to keep a lookout
